opensrp-opd/src/main/java/org/smartregister/opd/tasks/FetchRegistrationDataTask.java

```diff
@@ -0,0 +1,44 @@
+package org.smartregister.opd.tasks;
+
+import android.content.Context;
+import android.os.AsyncTask;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.smartregister.CoreLibrary;
+import org.smartregister.opd.utils.OpdConstants;
+import org.smartregister.opd.utils.OpdJsonFormUtils;
+import org.smartregister.opd.utils.OpdReverseJsonFormUtils;
+
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+import java.util.Map;
+
+public class FetchRegistrationDataTask extends AsyncTask<String, Void, String> {
+
+    private WeakReference<Context> contextWeakReference;
+    private OnTaskComplete onTaskComplete;
+
+    public FetchRegistrationDataTask(@NonNull WeakReference<Context> contextWeakReference, @NonNull OnTaskComplete onTaskComplete){
+        this.contextWeakReference = contextWeakReference;
+        this.onTaskComplete = onTaskComplete;
+    }
+
+    @Nullable
+    protected String doInBackground(String... params) {
+        Map<String, String> detailsMap = CoreLibrary.getInstance().context().detailsRepository().getAllDetailsForClient(params[0]);
+
+        detailsMap.put(OpdJsonFormUtils.OPENSRP_ID, detailsMap.get(OpdConstants.KEY.OPENSRP_ID));
+
+        return OpdReverseJsonFormUtils.prepareJsonEditOpdRegistrationForm(detailsMap, Arrays.asList(OpdJsonFormUtils.OPENSRP_ID, OpdConstants.JSON_FORM_KEY.BHT_ID), contextWeakReference.get());
+    }
+
+    protected void onPostExecute(@Nullable String client) {
+        onTaskComplete.onSuccess(client);
+    }
+
+    public interface OnTaskComplete {
```

Comment:
Can this be separate from this class, cause i think it can be reused by other Async Tasks.....What do you think?

Reply:
Nope, it can't. No need for it to be separate and it's not going to be reused for any other purpose. It is specifically for this DataTask. Do you have suggestions of where else it would be reused

Reply:
Just a thought why do you even need `onTaskComplete` the task can be performed in the `postExecute`

Reply:
Why move all the code to the AsyncTask? The AsyncTask is like an executor, all it does is that it runs whatever you want in the background. It is not supposed to have an specific functionality in it.

Reply:
It has `onPostExecute` to handle any result from the task that has been processed in the background